### PR TITLE
Impl: Remove preview channel on pr close

### PR DESCRIPTION
**Pull request by kozr**
# Description

Happy holidays!

Closes: #60

### [**DEMO/EXAMPLE**](https://github.com/kozr/action-hosting-deploy/pull/5)
The only difference between the demo branch in my fork and this branch is the .github/workflow/deploy-preview.yml. The main branch in my fork already has the preview template so I could test.

Although not completely done with what #60 was asking. I implemented the part where a preview channel would be closed after a PR closes (important to note that for a PR: closed != merged). There [isn't an way](https://github.community/t/trigger-workflow-only-on-pull-request-merge/17359) to detect when a PR is merged unless we detect it from the branch in which it is merged into.

Edit: Found out that "closed" will trigger on PR merge. Hooray, it is what #60 is asking for.

Feedbacks are always welcome.

New things:
**_removePreview_**
Mostly identical to the structure of deployPreview except when parsing the results and the argument.
An extra `--force` flag is added to skip cli prompts
```
  const removeDeployment = await execWithCredentials(
    ["hosting:channel:delete", channelId, "--force"],
    projectId,
    gacFilename
  );
```
I added a couple of tests that are similar to the ones for deployPreview

**_removeOnClose_**
A new flag for the GitHub action to enable this feature, defaults to false.
I documented the relevant information regarding the flag in README.md

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Worked on my forked repo with a new Firebase project

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kozr** [comment]
Hold on, let me fix the deploy preview yml

**Comment by napei** [comment]
> There [isn't an way](https://github.community/t/trigger-workflow-only-on-pull-request-merge/17359) to detect when a PR is merged unless we detect it from the branch in which it is merged into.

From that link I think you can infer that a PR has been "merged" regardless of if it's closed manually or merged? 
```yaml
github.event.pull_request.merged
```
and/or 
```yaml
- name: pull request
        if: github.event_name == 'pull_request' && github.event.action == 'closed' && github.event.pull_request.merged == true
```

I haven't tested anything, just making sure it's not missed 😄

**Comment by kozr** [comment]
I totally missed that. I will look into it, looks like it'll work though. Thanks @napei

**Comment by napei** [comment]
> I totally missed that. I will look into it, looks like it'll work though. Thanks @napei

Another extension for this behavior is for re-opening PRs if that's even detectable. That could be a separate issue alltogether and might already be implemented I'm just throwing ideas.

File: .github/workflows/deploy-preview.yml [hunk 1]
```diff
@@ -14,7 +14,9 @@
 
 name: Deploy Preview
 
-on: pull_request
+on:
+  pull_request:
+    types: [closed, synchronize, reopened, opened]
```
Comment (kozr) [inline .github/workflows/deploy-preview.yml:19]: The normal pull_request trigger already contains synchronize, reopened, and opened. So it works already!

File: bin/action.min.js [hunk 2]
```diff
@@ -11232,6 +11232,15 @@ async function deployPreview(gacFilename, deployConfig) {
   const deploymentResult = JSON.parse(deploymentText.trim());
   return deploymentResult;
 }
+async function removePreview(gacFilename, deployConfig) {
+  const {
+    projectId,
+    channelId
+  } = deployConfig;
+  const removeDeployment = await execWithCredentials(["hosting:channel:delete", channelId, "--force"], projectId, gacFilename);
```
Comment (prilutskiy) [inline bin/action.min.js:11240]: What if the `channelId` is autogenerated, and not set manually? When so, we can't know the channelId, which means we cannot delete it.
IMO, it makes sense to add a check that `removeOnClose` can only be used when `channelId` is set explicitly.

**Review by prilutskiy — COMMENTED** [review]
Is it possible to post a comment back to the PR, once the preview channel is closed? This kind of feedback would be very helpful, dev would then know what's going on.

For instance, semantic-release pushes back to merged PRs comments like "hey, changes from this PR has just been included in release X.Y.Z", which is cool.

**Comment by prilutskiy** [comment]
Great job! Can't wait for this to be released 😎

**Comment by kozr** [comment]
I'll be spending time on this tomorrow.

File: bin/action.min.js (hunk 2 above)
Comment (kozr) [reply bin/action.min.js:11240]: I think we have a function called `getChannelID` that detects that and returns the correct channelID whether it's manually set or auto-generated.

Comment (kozr) [reply bin/action.min.js:11240]: `  const channelId = getChannelId(configuredChannelId, context); ` exists to get the channelID somewhere in the `index.ts`file

**Comment by kozr** [comment]
> Is it possible to post a comment back to the PR, once the preview channel is closed? This kind of feedback would be very helpful, dev would then know what's going on.
> 
> For instance, semantic-release pushes back to merged PRs comments like "hey, changes from this PR has just been included in release X.Y.Z", which is cool.

I think this would be pretty cool. I might try to make another PR for it later, this PR is getting kind of big to review.

**Comment by kozr** [comment]
> > There [isn't an way](https://github.community/t/trigger-workflow-only-on-pull-request-merge/17359) to detect when a PR is merged unless we detect it from the branch in which it is merged into.
> 
> From that link I think you can infer that a PR has been "merged" regardless of if it's closed manually or merged?
> 
> ```yaml
> github.event.pull_request.merged
> ```
> 
> and/or
> 
> ```yaml
> - name: pull request
>         if: github.event_name == 'pull_request' && github.event.action == 'closed' && github.event.pull_request.merged == true
> ```
> 
> I haven't tested anything, just making sure it's not missed 😄

Actually, I tested and realized that merged PR counts as "closed", so the gh action would run anyway. Nice.

https://github.com/kozr/action-hosting-deploy/pull/6 
https://github.com/kozr/action-hosting-deploy/actions/runs/455240888

**Comment by kozr** [comment]
@jhuleatt thoughts?

**Comment by 0xdevalias** [comment]
Curious how progress on this is going. Would love to see this feature land!

**Comment by kevmo314** [comment]
> Curious how progress on this is going. Would love to see this feature land!

Likewise. Is this just waiting for review? We have a very high pull request velocity on our team and it's using up our channel quota.

**Comment by Guthers** [comment]
Bumping this was one of the first things I went looking for after I setup CI/CD. Great imitative taking this on!

**Comment by Seamoo13** [comment]
@kozr Any updates? I really need this on my project :sweat_smile:

**Comment by rvargas** [comment]
Any update?

**Comment by radvansky-tomas** [comment]
Why this is not merged yet?

**Comment by razine-bensari** [comment]
it has been 3 years... I dont understand this is a really nice feature!

**Comment by kozr** [comment]
Better to rewrite it than to fix conflicts with 3 years worth of changes.

**Pull request closed without merging**